### ISSUE TITLE
Implementing ProgressDialog with WPF in a ManagedUI

### ISSUE THREAD

**Issue by GarrenMiller** [oleg-shilo/wixsharp#1034]
I've written an installer with WixSharp using a ManagedUI, and am currently using the predefined WinForms dialogs.  In particular, my issue is with the ProgressDialog.  I noticed that at higher scaling factors, the WinForms ProgressDialog is misaligned (as mentioned in issue #974).  From the discussion on #974, I gathered that porting to WPF was the correct solution and that WixSharp does not yet have a predefined set of WPF dialogs.  

I decided to try my hand at writing my own WPF progress dialog, so I followed this example: https://github.com/oleg-shilo/wixsharp/tree/master/Source/src/WixSharp.Samples/Wix%23%20Samples/Managed%20Setup/CustomUI.WPF

I am able to create a WPF dialog and add it to the installer dialog sequence just fine.  The issue comes when I try to create an override for the OnExecuteComplete() and OnProgress() methods.  I've included the relevant part of ProgressDialogWpf.xaml.cs used to define my dialog.  As you can see, my ProgressDialogWpf class implements IWpfDialog and to my understanding this should allow me to define overrides for the two methods I mentioned.

```
namespace WindowsInstaller.Dialogs
{
    public partial class ProgressDialogWpf: IWpfDialog
    {
        public ProgressDialogWpf()
        {
            InitializeComponent();
            DataContext = this;
        }

        public void Init()
        {
            Host?.Shell.StartExecute();
        }

        public override void OnProgress(int progressPercentage)
        {
            MessageBox.Show("Progress changed");
        }
        
        public override void OnExecuteComplete()
        {
            MessageBox.Show("Execution complete");
            Host?.Shell.GoNext();
        }
        ...
    }
}
```

Notice that I'm calling Shell.StartExecute() inside Init().  This works perfectly as I intended, and installation works just fine.  The installer runs to completion and all the files I expect to be present are there.  Everything is fine from an installation standpoint.  However, I cannot get either OnProgress() or OnExecuteComplete() to work the way they do in the WinForms version of the ProgressDialog.  I resorted to adding a MessageBox.Show() to each of them to see if they were even being called, and no message boxes are ever shown.  As far as I can tell, neither of these methods are getting called for the ProgressDialogWpf that I have created.

I tried the MessageBox.Show() trick with the tried-and-true WinForms ProgressDialog, and it worked like I would expect it to (i.e. the message boxes were shown when the methods were called).  

It is quite possible that there is something I do not understand about how this process works, but as far as I know what I'm trying to do should work.  I would appreciate any input or suggestions as to why what I'm trying to do isn't working.

**Comment by oleg-shilo** [comment]
Sounds like a defect. Will have a look.
Txs

**Comment by oleg-shilo** [comment]
Indeed the chain of cascading events was broken in the hosting layer.

Fixed now. The fix is available as v1.18.3.0-pre NuGet package. You will only need WixSharp.WPF package.

Sorry for the delay.

**Comment by GarrenMiller** [comment]
Excellent.  I will try to clear some time at some point to work on redesigning the default dialogs in WPF when I get a chance.

**Comment by GarrenMiller** [comment]
@oleg-shilo I spent some time yesterday and this morning working on a WPF version of the ProgressDialog.  OnProgress() and OnExecuteComplete() now function as expected, which is most of what I needed.  Thanks!

I am still having trouble getting ProcessMessage() to work at all.  I've implemented IProcessDialog, but it seems like my override never gets called.  This isn't a huge problem since the progress bar still works visually and can demonstrate to the user that the installation is making progress.  I just cannot provide a status message above the progress bar.

Also, a bit unrelated, but I am having trouble getting the WPF dialogs to use the localized UI strings.  For instance, when recreating the ProgressDialog I created a `<TextBlock Text="[ProgressDlgTextInstalling]"/>`.  I also tried `<TextBlock>[ProgressDlgTextInstalling]</TextBlock>`.  Both of these result in the text block just having the string `{\WixUI_Font_Title}Installing [ProductName]` instead of the desired string.  Is there something I need to do to allow the WPF dialogs to use the localized UI strings like the WinForms do?  I wasn't sure if I should open a separate issue for this or not, so let me know if I should do so.

**Comment by oleg-shilo** [comment]
Just released v1.20.0.0.

It delivers custom complete WPF UI:
https://github.com/oleg-shilo/wixsharp/tree/v1.20.0.0/Source/src/WixSharp.Samples/Wix%23%20Samples/Managed%20Setup/CustomUI.WPF.Sequence

The following line does the trick
```C#
project.ManagedUI = ManagedUI.DefaultWpf;
```

**Comment by oleg-shilo** [comment]
Corresponding VS2022 template has been published:
https://github.com/oleg-shilo/wixsharp/wiki/Deployment-scenarios#managed-wpf-ui---complete-ui
